### DISCUSSION TITLE
Weird behaviour when porting a book exercise to Silk.Net

### DISCUSSION THREAD

**Discussion by lehmamic** [dotnet/Silk.NET#1046] [Q&A]
Hi guys,

I'm reading the book Game Programming in C++ from Madhav Sanjay. I did the exercises in CPP and am now porting it to Silk.Net. Rendering Sprites works so far, but I have the problem, that my Qube does not get rendered properly.

I have no clue what it could be. Just some background info:
- The book uses row based vectors in a left-handed coord system.
- I'm trying to port the code 1:1, or at least as much as possible.

Here is my code:
https://github.com/lehmamic/GameProgrammingExercises

I'm to of ideas what it could be, I would appreciate your help very much, thanks.

**Comment by HurricanKai** [comment]
Hey, it's pretty difficult to tell what's wrong like this.
At a first guess I'd think something when wrong when you converted your maths.
C#/.NET matrices are commonly stored in row-major order, so maybe that's where it went wrong?

**Reply by lehmamic** [reply]
I set the transpose flag in the shader to true:
```

_gl.UniformMatrix4(location, 1, true, (float*) &value);
```

And calculate the position as following:
```
gl_Position = pos * uWorldTransform * uViewProj;
```

That should fit for row-major, correct? At least the Sprites work well with that.

One difference to the Sprites is the usage of Querternions, could it be that there is some difference to the CPP exercises?

**Reply by HurricanKai** [reply]
It's possible there's some difference how quaternions are handled. To be honest I have next to no idea about quaternion maths 😅

**Reply by lehmamic** [reply]
I found it. It was the initialization of the Quaternions.

The book does it this way:
```cs
float scalar = Scalar.Sin(angle / 2.0f);
var x = axis.X * scalar;
var y = axis.Y * scalar;
var z = axis.Z * scalar;
var w = Scalar.Cos(angle / 2.0f);

return new Quaternion<float>(x, y, z, w);
```

Silk.Net does it this way:
```cs
public Quaternion(Vector3D<T> vectorPart, T scalarPart)
{
  this.X = vectorPart.X;
  this.Y = vectorPart.Y;
  this.Z = vectorPart.Z;
  this.W = scalarPart;
}
```

It looks good now on Windows, but not on Mac M1, maybe its bug in OpenGL. The Cube example looks right dough....

Something else I detected:
It works when I hardcode the index array before I but it into the buffer object, but when I pre initialize the index array an fill it up by loaded values from json it gets corrupted somewhere. When I check both array, they are exactly equal....

Loaded from json:
```cs
  var indices = new uint[raw.Indices.Length * 3];
  for (int i = 0; i < raw.Indices.Length; i++)
  {
      var index = raw.Indices[i];
      if (index is null || index.Length != 3)
      {
          throw new MeshException($"Invalid indices for {fileName}.");
      }
  
      for (int j = 0; j < index.Length; j++)
      {
          indices[i * 3 + j] = index[j];
      }
  }
```

Manually initialized (its a cube):
```cs
var indices2 = new uint[]
{
    2, 1, 0,
    3, 9, 8,
    4, 11, 10,
    5, 11, 12,
    6, 14, 13,
    7, 14, 15,
    18, 17, 16,
    19, 17, 18,
    22, 21, 20,
    23, 21, 22,
    26, 25, 24,
    27, 25, 26,
};
```

I use the buffer object from the examples. The values in both arrays are equal. What could that be?
